Scalebar

PhilipeRLeal opened the pull request:
# Issue #490

This Pull Requests closes #490

## Rationale

In respect to the discussed in  #490, here is an implementation of a scalebar for geoaxes.

The algorithm allows one to import the scalebar function directly from cartopy, or be directly called from the Goeaxes.


## Implications

None applicable

## Limitations:

So far, I have only been capable of inserting the scalebar without a background. 

Here is an example for using it:

[
![geoaxes plots with respective scalebar  for multiple projections](https://user-images.githubusercontent.com/32545466/107090562-27523900-67df-11eb-81f3-4202b8dd11ea.png)
](url)

## TODO:

Add some background for the scalebar

### Code for generating the above examples:


```

import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from cartopy.mpl.scalebar import fancy_scalebar


def test_scalebar():
    """Test the scalebar"""

    fig, axes = plt.subplots(1, 2,
                             subplot_kw={'projection':
                                         ccrs.Mercator()})

    projections = [ccrs.Mercator(), ccrs.PlateCarree()]

    axes = axes.ravel()
    try:
        for proj, ax in zip(projections, axes):
            ax.projection = proj
            fancy_scalebar(ax,
                           location=(0.5, 0.2),
                           length=1000_000,
                           unit_name='km',
                           max_stripes=5,
                           fontsize=8,
                           dy=0.05)

            ax.gridlines(draw_labels=True)
            ax.stock_img()
            ax.coastlines()
        condition = True
    except BaseException as err:
        print(err)
        condition = False

    assert(condition)


def test_scalebar_within_geoaxes():
    """Test scalebat within the geoaxes"""

    fig, axes = plt.subplots(1, 2,
                             subplot_kw={'projection':
                                         ccrs.Mercator()})

    projections = [ccrs.Mercator(),
                   ccrs.PlateCarree()]

    axes = axes.ravel()
    try:
        for proj, ax in zip(projections, axes):
            ax.projection = proj

            ax.set_extent([-60, -35, -40, 10])
            ax.gridlines(draw_labels=True)
            ax.add_scalebar(location=(0.5, 0.5),
                            length=250_000,
                            dy=5,
                            max_stripes=3)
            ax.stock_img()
            ax.coastlines()

        condition = True
    except BaseException as err:
        print(err)
        condition = False

    assert(condition)


if '__main__' == __name__:
    test_scalebar()
    test_scalebar_within_geoaxes()
```

adrien-berchet committed:
This is a merge of the prior developers git branch and the new functionalities here provided. The prior developers branch had several commits which were aggregated into
this single 1st commit.

Add a warning and comments

This is the commit that antecedes all added new commits in respect to the pull request number #1693 from: https://github.com/SciTools/cartopy/pull/1693

PhilipeRLeal committed:
changes in the cartopy.mpl.gridliner class. It has methods for easily setting the number of ticks, and changing the decimal separators in the ticklabels

PhilipeRLeal committed:
There were changes in the LOGITUDE_FORMATTER and the LATITUDE_FORMATTER. Now, the user can directly change the hemispheres (north-south and west-east) ticklabels ('N', 'S', 'W', 'E') from the gridliner instance. Also, new functionalities have been added to the gridliner

PhilipeRLeal committed:
scalebar has been added, as well as its test suit

PhilipeRLeal committed:
contributor list has been updated

PhilipeRLeal committed:
requirements documentation has been updated

PhilipeRLeal committed:
an example of the scalebar functionality has been added

PhilipeRLeal committed:
importation of the formatter module has been fixed

PhilipeRLeal committed:
formatters.py, gridliner.py, scalebar.py and scalebar_example.py have been changed to be conform with flake8

PhilipeRLeal committed:
changes to the scalebar_example.py

PhilipeRLeal committed:
the geopy requirement is no longer needed. Scalebar module is running ok

PhilipeRLeal committed:
scalebar example has been updated

PhilipeRLeal committed:
the package can now create a scalebar for practically any cartopy's projection

PhilipeRLeal committed:
test suit for scalebar has been updated

PhilipeRLeal committed:
test suit has been updated

PhilipeRLeal committed:
scalebar can now be called directly from the geoaxes. Also, the test for the scalebar has been added

PhilipeRLeal committed:
The lib has been updated in accordance to PEP8 format. Besides, the scalebar has been updated. Now, the fancy_scalebar has a proper documentation. Also, the geoaxes can directly create a scalebar

PhilipeRLeal committed:
test_scalebar has been updated

PhilipeRLeal committed:
formatters, scalebar and test_scalebar modules have had their copyright added, following the cartopy contributors' standard'

PhilipeRLeal committed:
There were changes in the formatter.py, gridliner.py, and the respective examples and tests. \n The changes here applied allow an easy assess and respective setting of the geoaxes.griline's ticklabels. The settings that can now be changed by the user are the following: \n a) the decimal separator (from dot '.' to any other string (i.e. from dot to comma ','); \n b) the number of decimal places after the coordinate (i.e.: {0:3f}); \n c) the labels of the east-west hemispheres ('W', 'E') - to any other set of labels (i.e.: 'Oeste', 'Leste'); \n d) the labels of the north-south hemispheres ('N', 'S') - to any other set of labels (i.e.: 'Norte', 'Sul'>.

PhilipeRLeal committed:
first commit of the branch solely_scalebar. In this first comit, the cartopy's original master branch has been updated by the addition of the scalebar method in the geoaxes. A test and an example files of the scalebar method are also provided

PhilipeRLeal committed:
scalebar_usage_examples.py has been updated in accordance to pep8 standards

PhilipeRLeal committed:
small changes in the comments from the scalebar_example

PhilipeRLeal committed:
first commit of the gridline ticklabel branch

PhilipeRLeal committed:
griliner test has been updated

PhilipeRLeal committed:
there were updates in the LicenseHeaders of the tests and the examples of the gridliner module and submodules

PhilipeRLeal committed:
test_gridliner.py has been updated. Now, error messages are returned for better follow up. Furthermore, the new figure tests have the ImageTesting decorator in its functions.

PhilipeRLeal committed:
small changes in variables given long name problems from PEP8

PhilipeRLeal committed:
both branches have been properly merged. COnflict merging has been solved

PhilipeRLeal committed:
The scalebar is now inserted to the geoaxes through the matplotlib's offset subclasses. TODO: there is still some problems regarding how to add the ticklabels into the scalebar's ticklabels...

PhilipeRLeal committed:
the scalebar test module has been updated

PhilipeRLeal committed:
stickler problems were solved

PhilipeRLeal committed:
Merge branch 'scalebar' into master

PhilipeRLeal committed:
test_scalebar.py has been updated. The test was not passing beforehand

PhilipeRLeal committed:
stickler hsa been solved

PhilipeRLeal committed:
Update test_scalebar.py

The angle attribute has been removed from the scalebar function

PhilipeRLeal committed:
Update test_scalebar.py

Figure closure functions were deleted.

PhilipeRLeal committed:
Merge branch 'master' of https://github.com/PhilipeRLeal/cartopy into HEAD

PhilipeRLeal committed:
Merge branch 'solely_scalebar' of https://github.com/PhilipeRLeal/cartopy into HEAD

PhilipeRLeal committed:
Merge branch 'master' into scalebar

PhilipeRLeal committed:
scalebar is running OK. It can be plotted against several projections

PhilipeRLeal committed:
small updates in the ruler unit font formatter

PhilipeRLeal committed:
PEP8 standardized

PhilipeRLeal committed:
Update test_scalebar.py

Update in the importation of the scalebar aad_scalebar.

PhilipeRLeal committed:
Create test_scalebar.py

Update in the importation of the scalebar aad_scalebar.

PhilipeRLeal committed:
Update test_scalebar.py

corrections on the parameters of the function add_scalerbar.

PhilipeRLeal committed:
Update test_scalebar.py

Typo fix in inline call of the scalebar

PhilipeRLeal commented:
Dear developers, now that the scalebar Issue is basically solved, should I close this Issue?

Sincerely,

greglucas commented:
Your code isn't merged in yet, so don't close any of the issues yet. If you put "closes #XX" with a link to the issue you want to close in your PR request it will automatically close the linked issues when someone merges the PR.

PhilipeRLeal commented:
Dear @greglucas, thank you for the information. I have updated the body of this Pull Request adding the term:"closes #490". Therefore, I believe that once this Pull Request os merged, the Issue 490 will be then closed. Corret?

Sincerely,

PhilipeRLeal committed:
Update scalebar.py

description of the add_scalebar function was updated.

PhilipeRLeal committed:
Merge branch 'master' into scalebar

PhilipeRLeal committed:
lint errors solved

PhilipeRLeal committed:
Merge branch 'master' into scalebar

PhilipeRLeal committed:
Merge branch 'master' into scalebar

QuLogic commented:
Running PEP8 (or whatever formatter you used) on _everything_ is very confusing, as it muddles what it is you actually changed here. You also seem to have mixed in your other PR here, so I really don't know what to look at.

PhilipeRLeal committed:
Merge branch 'master' into scalebar

PhilipeRLeal committed:
Update scalebar.py

numpy array's base method has been deleted within the "_point_along_line" function

PhilipeRLeal committed:
Update scalebar.py

PhilipeRLeal committed:
Update scalebar.py

numpy "base" method was dropped from the geodesic.inverse results.

PhilipeRLeal committed:
Update scalebar.py

Typo solved from the TextArea

PhilipeRLeal committed:
Update test_gridliner.py

The Old "AttributeError: module 'cartopy.crs' has no attribute 'PROJ4_VERSION'
_________________ ERROR collecting tests/mpl/test_scalebar.py _________________
" has been solved in this Commit

PhilipeRLeal committed:
Update test_gridliner.py

Spacing error solved

PhilipeRLeal committed:
Update scalebar.py

typo solved

PhilipeRLeal committed:
Update scalebar.py

minimumdescent=True is now deprecated within matplotlib

PhilipeRLeal committed:
Update test_gridliner.py

PhilipeRLeal committed:
Update gridliner.py

PhilipeRLeal committed:
Delete gridliner.py

PhilipeRLeal committed:
Create gridliner.py

Copied from the cartopy's master branch

PhilipeRLeal committed:
Delete formatters.py

PhilipeRLeal committed:
Update gridliner.py

It is now exactly equal to cartopy's master branch

PhilipeRLeal commented:
> Running PEP8 (or whatever formatter you used) on _everything_ is very confusing, as it muddles what it is you actually changed here. You also seem to have mixed in your other PR here, so I really don't know what to look at.

@QuLogic, I have tried to clean all this confusion into this new commit. See if it helps.

scollis commented:
Just want to add this is exactly the feature I am looking for! Need any help getting this merged?

PhilipeRLeal commented:
> Just want to add this is exactly the feature I am looking for! Need any help getting this merged?

Yes, please. I have done all I could do to facilitate the merge. Nevertheless, the time I updated this Pull Request, there were several LINT errors that would not allow a Merge. Therefore, I had to fix them all (even the ones that were not caused by this branch). As a consequence, it seems that this branch commit's history is now denser than the reviewers expected.

Let me know if I can be of any service.

Sincerely,

drewskoots commented:
Any update on the progress with this? If not, is there something a complete beginner to this particular toolbox could do to move it along? Would love to see this feature implemented!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SciTools/cartopy?pullRequest=1728) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SciTools/cartopy?pullRequest=1728) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: adrien-berchet<br/>:x: PhilipeRLeal<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SciTools/cartopy?pullRequest=1728) it.</sub>
